### PR TITLE
Feature: card option updates

### LINKED ISSUE #2306: Implement card option updates from UIDS
In UIDS, events are just going to be one more thing that can be created with a card component. This means that cards will now support the variations that are currently limited to events. To support this change, we need to make the following changes
* Add LB style options for:
  - Image styles: rectangular (default) or round
  - Image size (when a card is in horizontal orientation): small, medium (default), or large
  - Image position (when a card is in horizontal orientation): left (default), or right, or above.
* Solve for programmatically setting the card image style to be square instead of widescreen when the round option is selected above.
* Show/hide the image size and position options as appropriate when the horizontal variant is selected.
* Add ability to add more background colors to card like `light` `dark` and `gray`.

### PR DESCRIPTION
**Pull request by bspeare**
Resolves: https://github.com/uiowa/uiowa/issues/2306

# How to test
*If you get an error like `Error: File to import not found or unreadable`, you need to run the following commands:*
- `rm -rf ./node_modules`
- `yarn cache clean`

Steps:
- `blt frontend`
- **Test update hook:** test on `v3` sites with vertical card rows (https://uiowa.edu/academics/areas-study, https://uiowa.edu/academics/leading-programs, https://uiowa.edu/academics/career-preparation) and horizontal card stacks (https://uiowa.edu/admissions, https://coronavirus.uiowa.edu/health). 
- **Existing card implementations:**  test on `v2` and `v3` article lists, people lists, featured content lists, and cards. 
- **New display options:** test all the new card display options using the new display options. Check to make sure that hiding and showing of the image size column works as described below in 'Programmatic updates' and that, if you do trick the form into allowing you to enter a forbidden combination, it is not saved.

# Updates 

## Programmatic updates

- Programmatically setting the card image style to be square instead of widescreen when the circle/round 1:1 option is selected
- Hiding the image size option when the stacked image option is selected AND either the square or widescreen image format is selected
- Form submission logic to reset image size to large if the user selects a different size when the option is shown, but then changes other options that would no longer allow it.

## Card options

**Image position**  - select one
- Stacked (default)
- Left - this and the next option replace the `horizontal` option.
- Right

**Image format** - select one
- Widescreen (16:9) (default)
- Square
- Circle

**Image size** - select one
- Small 
- Medium
- Large (default)

**Background color** - select one
- Default is still a white background
- Only `gray` currently is an option

**General Styles** - select any
- Border

### CODE REVIEW

**Comment by GaryRidgway** [comment]
Some notes about what I am seeing so far: 

Currently buttons are overflowing the container on each section in the horizontal view.
Screen Shot 2020-10-26 at 2 27 11 PM

This can be solved by removing the `width: max-content;` rule on the button.

Second the grey and yellow BGs have incorrect padding on buttons:
Screen Shot 2020-10-26 at 2 27 21 PM
Screen Shot 2020-10-26 at 2 27 16 PM

and the yellow and Brain have unseeable button borders on the background color: 
Screen Shot 2020-10-26 at 2 27 27 PM

The images are also able to get VERY small, and for editors even invisible. Maybe setting a minimum width?
Screen Shot 2020-10-26 at 2 30 34 PM

Also, the editor is a bit misleading, as it shows a white background instead of the actually displayed background. Here is an example with a yellow BG, but it shows a white BG on the layout builder page:
Screen Shot 2020-10-26 at 2 42 25 PM

**Comment by bspeare** [comment]
@GaryRidgway thanks! I've started a list below - can you add any other additional observations from your screenshots to this list?

- [ ] Button/color border styles with different background colors
- [ ] Button overflow on horizontal cards 
- [ ] Admin screen appearance - remove specific card background styles from `admin.scss` 
- [ ] Square image not becoming edge to edge on cards with explicit BG colors: gray, widescreen, stacked, center, large. 
- [x] Adjust border color on circle to match `card--enclosed` border color. 
- [ ] Revisit padding for different image styles with background colors: gray, circle, large.
- [x] Adjust update hook for `card--horizontal` usage to use: widescreen, left, left (default), medium. 
- [ ] Adjust image styles to output smaller images if `small` or `medium` are selected. 
- [x] Increase `small` image size slightly.

**Comment by pyrello** [comment]
The update hook now accounts for cards that are currently horizontal.

When testing this on the default site (which conveniently has both vertical and horizontal cards), it looks like there are some issues with the display of the horizontal card in default format:
Screen Shot 2020-10-27 at 7 50 06 AM

Specifically, the image seems to be bottom aligned to the container, which seems wrong.

**Comment by pyrello** [comment]
Updated the description of style changes to reflect how the project has progressed.

**Comment by pyrello** [comment]
Nevermind my previous comment about the horizontal card.

File: yarn.lock [hunk 1]
```diff
@@ -932,9 +932,15 @@
   resolved "https://registry.yarnpkg.com/@types/q/-/q-1.5.4.tgz#15925414e0ad2cd765bfef58842f7e26a7accb24"
   integrity sha512-1HcDas8SEj4z1Wc696tH56G8OlRaH/sqZOynNNB+HF0WOeXPaxTtbYzJY2oEfiUxjSKjhCKr+MvR7dCHcEelug==
 
-"@uiowa/uids@uiowa/uids#3.0.7":
+<<<<<<< HEAD
```
Comment (joewhitsitt) [inline yarn.lock]: Looks like bad merge remains

Comment (bspeare) [reply yarn.lock]: Thanks!

**Comment by dnkemper** [comment]
I can start testing this!

**Comment by joewhitsitt** [comment]
I plan to test soon, but would docroot/sites/ighn.international.uiowa.edu/modules/ighn_core/templates/node--ighn-person--teaser.html.twig card be in-scope?

**Comment by dnkemper** [comment]
Unrelated: Does it matter that cards added underneath cards become merged with the card above it? Did that even make sense? Haha
Screen Shot 2020-10-30 at 2 27 53 PM

### **testing on default site
question: When image position = stacked, and image style = circle, you have the ability to choose large, medium and small sizes for the image... BUT can't with widescreen image style or square image style. Does that matter?

### **tested:**
- Image position: stacked, Image format: circle, Content alignment: center, Image size: small, medium, large
- Image position: stacked, Image format: circle, Content alignment: left, Image size: small, medium, large
- Image position: stacked, Image format: widescreen & square, Content alignment: left, Image size: none
- Image position: stacked, Image format: widescreen & square, Content alignment: center, Image size: none
- Image position: left, Image format: widescreen, Content alignment: left, Image size: small, medium, large
- Image position: left, Image format: square, Content alignment: left, Image size: small, medium, large
- Image position: left, Image format: circle, Content alignment: left, Image size: small, medium, large
- Image position: left, Image format: widescreen, Content alignment: center, Image size: small, medium, large
- Image position: left, Image format: square, Content alignment: center, Image size: small, medium, large
- Image position: left, Image format: circle, Content alignment: center, Image size: small, medium, large
- Image position: right, Image format: widescreen, Content alignment: left, Image size: small, medium, large
- Image position: right, Image format: square, Content alignment: left, Image size: small, medium, large
- Image position: right, Image format: circle, Content alignment: left, Image size: small, medium, large
- Image position: right, Image format: widescreen, Content alignment: center, Image size: small, medium, large
- Image position: right, Image format: square, Content alignment: center, Image size: small, medium, large
- Image position: right, Image format: circle, Content alignment: center, Image size: small, medium, large

**Comment by bspeare** [comment]
@joewhitsitt Yes, you might need to change `card_author` to `card_subtitle` in that template.  I can make that fix and test it if that would help.

**Comment by joewhitsitt** [comment]
@bspeare thanks. swapped it. still hoping to start testing today...

**Comment by bspeare** [comment]
@dnkemper good point about stacking the cards.   We will probably not solve that in this issue but we should put in a rule to add space.     You are correct, you can't change the image size for `stacked` `widscreen` or `square` but you can for`circle`.

**Comment by dnkemper** [comment]
Thank you, @bspeare!

**Comment by GaryRidgway** [comment]
Testing the Hooks right now. stacked cards seem to be working but horizontal cards did not. I am getting stacked images on those. 
This is Live Admissions
Screen Shot 2020-10-30 at 3 24 15 PM
And its config
Screen Shot 2020-10-30 at 3 24 50 PM
And this is My local one
Screen Shot 2020-10-30 at 3 24 26 PM
And its config
Screen Shot 2020-10-30 at 3 24 58 PM

**Comment by pyrello** [comment]
@GaryRidgway Thanks! I'll revisit the update code. I have not tested it against the admissions site or even in the last few days.

**Comment by GaryRidgway** [comment]
@pyrello  Totally! I also tested on https://coronavirus.uiowa.edu/health and got the same results

**Comment by cory-skeers** [comment]
Hrm... https://coronavirus.uiowa.edu/health update hook worked for me on the horizontal cards (sets to left-aligned image, image-size small). Haven't tested admissions. Nothing's jumping out at me as suspect, but I can try to dig around too, and test it on admissions.

**Comment by pyrello** [comment]
@cory-skeers The only thing I can think of that would be out of place is that I didn't update a machine name for a LB style after updating it.

**Comment by cory-skeers** [comment]
Checked it on https://uiowa.edu/admissions, and it worked correctly for me.

Also noticed in the screenshot that it has the image size field showing alongside image: stacked and widescreen. @pyrello I haven't looked at the code that handles that logic, but that might be a clue to where there could be a mismatch?

**Comment by cory-skeers** [comment]
@GaryRidgway Are you able to select Left or Right image placement? I'm wondering if maybe it's a config bug, and the block is being set with left image alignment, but the style isn't in place. Then the form might be displaying `Stacked (default)` if it's the only option and ` - none - ` has been removed, even if the value behind it is `card_media_position_left`.

**Comment by GaryRidgway** [comment]
@cory-skeers The three options are there, but actually even when I select left positioned it doesn't change anything. maybe this is a big on my end? I'm not really sure how to reset my config or anything though?

**Comment by GaryRidgway** [comment]
Oh wait, ugh, I think I didnt load the new styles, my apologies all. Let me try again after recompiling the styles.

**Comment by GaryRidgway** [comment]
@cory-skeers & @pyrello my apologies, I hadn't compiled styles. It is working fine now, the images are properly positioned and sized. Sorry for the confusion, I was too excited to test!

**Comment by pyrello** [comment]
@GaryRidgway I did this exact thing so many times over the course of this PR :|

**Comment by joewhitsitt** [comment]
Am I the only one encountering:
`Error: File to import not found or unreadable: components/card/_card-mixins.scss` on several components?

**Comment by pyrello** [comment]
@joewhitsitt This has been a problem recently, where yarn seems to be not correctly updating UIDS to the latest version. I updated the testing instructions with the additional steps that may be necessary to get that fixed.

**Review by cory-skeers — COMMENTED** [review]
Quick rundown of testing I've done, and that is currently in a passing state:
- Update hook on vertical and horizontal cards
- Display options (generally) and form logic
- V2 cards, article list, people list, featured content list
- V3 cards in various combinations, implementations

Haven't tested the various v3 list implementations.

**Comment by GaryRidgway** [comment]
I have utilized hawkeyemarchingband.uiowa.edu to test v2 article lists, people lists, featured content lists, and cards. Everything seems to be working functionally. The only thing I have noticed is some possible style discrepancies on featured content lists, but this may be due to things outside of the scope of this issue. There are no borders between listings like there are for other teaser lists, and the images are a little odd because there are interspersed widescreen and square images between people and articles.

Screen Shot 2020-11-02 at 10 06 44 AM

I'm not sure if this is set per card, person, page, or article. If this is not in scope please ignore!

**Comment by joewhitsitt** [comment]
> images are a little odd because there are interspersed widescreen and square images between people and articles

this is "as designed" for now as odd as it looks together.

**Review by GaryRidgway — APPROVED** [review]
I have taken a look at the various v3 lists and they look good. That combined with the rest of the info in this issue and the testing of everyone here, I think this is ready to go.

**Comment by joewhitsitt** [comment]
late to the party...

ighn.international's card implementation seems to be okay.

nothing appears to be broken in v2. 🤞
